### ISSUE TITLE
Backmerging Module Federation v2

### ISSUE THREAD

**Issue by ScriptedAlchemy** [webpack/webpack#18809]
## Feature request
This is a tracking issue for various pieces of work to backmerge some of the advances in module federation 2 back into webpack core. 

**What is the expected behavior?**

**What is motivation or use case for adding/changing the behavior?**
I want to sync these changes into rspack and want to reduce the number of patches i must maintain in my plugin that fix "bugs in webpack" 

**How should this be implemented in your opinion?**
TBD

**Are you willing to work on this yourself?**
yes, but I will need contributor help - i have taken it about as far as i know how to.

**Comment by einarq** [comment]
Will there at some point be an announcement about any potential MF 2 enhancements we can enable via Webpack?

I dont see us moving away from webpack anytime soon (a lot of apps deoend on it internally), so a bit confused on how to leverage new features coming from MF 2.

**Comment by ScriptedAlchemy** [comment]
V2 provides a webpack,rspack,vite,rollup plugin. Don't need to move away from webpack. Just stop using webpacks built in plugin. V2 is just a npm package. 

If we do sync the capabilities back, webpack will still have to depend on my central runtime since the build plugin is just a basic wrapper that implements my central runtime.

**Comment by einarq** [comment]
Is there a migration guide somewhere?Can I gradually move apps over, or do they all have to switch at once?Is the api of the new plugin the same? On 8 Nov 2024, at 21:39, Zack Jackson ***@***.***> wrote:﻿
V2 provides a webpack,rspack,vite,rollup plugin. Don't need to move away from webpack. Just stop using webpacks built in plugin. V2 is just a npm package.
If we do sync the capabilities back, webpack will still have to depend on my central runtime since the build plugin is just a basic wrapper that implements my central runtime.

**Comment by ScriptedAlchemy** [comment]
Just just new plugin. Start with host apps since v2 remote may no longer be compatible with v1 host. We stopped guaranteeing full interop about 5 months ago. V2 can consume v1 don't know about the other way around tho. 

No migration guide needed. If you use hacks etc you might have to replace the hacks with the now official mechanisms we provide in v2. Like dynamic remotes or promise new promise or using runtime plugins.

**Comment by einarq** [comment]
We have about 20 apps, as part of an enterprise platform, that share
federated modules.

Most just consume, 1 only expose, and some both consume and expose.
Sounds like it will be hard to gradually upgrade then.
I guess it will have to be a 3 step approach.
1. Upgrade the ones that only consume
2. Upgrade the ones that both consume and expose (will have to be deployed
together, otherwise things might break)
3. Upgrade the one that only exposes.

Right?

Btw, we should use @module-federation/enhanced now, right? Should a v1 be
released, or do you still consider MF v2 to be in pre-release stage?

**Comment by ScriptedAlchemy** [comment]
It's stable. Remember, ByteDance has used v2 internally for over two years, and we run some of the largest applications in the world. If you don't know what ByteDance is, Google some of the apps we own.
It's still pretty v1 because we are still extracting features and APIs from the company to send back to OSS. 

We have over 10,000 remotes active right now.

**Comment by einarq** [comment]
Have no issue trusting the quality of the project. Just thinking a v1 would be good to signal to everyone that it’s ready. Always feels a bit off to have a 0.x version in prod code. What was the strategy for upgrading all those bytedance apps to v2? Or were they never on v1? Do you agree that I would need something like the strategy I outlined earlier?Thanks On 11 Nov 2024, at 18:49, Zack Jackson ***@***.***> wrote:﻿
It's stable. Remember, ByteDance has used v2 internally for over two years, and we run some of the largest applications in the world. If you don't know what ByteDance is, Google some of the apps we own.
It's still pretty v1 because we are still extracting features and APIs from the company to send back to OSS.
We have over 10,000 remotes active right now.

**Comment by ScriptedAlchemy** [comment]
Both still have the get init primitive. So technically the base v1 capability should work. At byte v1 was like 3 years ago so I didn't work here then haha. 

I've only been around for v2. We keep it pre zero just because we are still moving new apis from byte into oss. 

I think v1 ill ship next year tho. Which means we just settle on internal apis and stuff.

**Comment by einarq** [comment]
Whats the long term plan re "backmerging" into core weback? 

Will all MF 2 capabilities eventually make it into the webpack plugin, and therefore we might as well wait for that? Just asking due to the name of this issue.

**Comment by ScriptedAlchemy** [comment]
Unknown. I want to reduce the number of patches I applied to webpack via my plugin.
Syncing with webpack is not a priority due to output velocity requirements at bytedance. But in the process of building v2, i did patch many problems in webpack, i want to move those out of my repo. 

I want to maintain central control over federation, since I do not have merge and release powers over webpack - I will not move anything into this repo that may hinder iteration speeds. Either way, webpack will have to add my packages as a dependency, it could re-export my classes if it wanted to. 

My runtime is released in tandem with my plugins during updates. You can wait to see what happens, but it will take you a few hours to migrate, or you can wait 6-12 months for something that might not happen. 

Either way, you'll still be installing module-federation/enhanced. 

I need to be able to release new versions of the system within an SLA of 15 minutes; getting blocked by release cycles is a non-negotiable. 

The challenge we have seen is Bytedance engineering moves faster than any community based project. Federation is a core technology that powers most of the thousands of apps we own. 

Backmerging parts is a spare time effort of which only i do. 

Just migrate.

**Comment by einarq** [comment]
I hear ya :)Is there a migration guide somewhere?Thanks On 12 Nov 2024, at 20:13, Zack Jackson ***@***.***> wrote:﻿
Unknown. I want to reduce the number of patches I applied to webpack via my plugin.
Syncing with webpack is not a priority due to output velocity requirements at bytedance. But in the process of building v2, i did patch many problems in webpack, i want to move those out of my repo.
I want to maintain central control over federation, since I do not have merge and release powers over webpack - I will not move anything into this repo that may hinder iteration speeds. Either way, webpack will have to add my packages as a dependency, it could re-export my classes if it wanted to.
My runtime is released in tandem with my plugins during updates. You can wait to see what happens, but it will take you a few hours to migrate, or you can wait 6-12 months for something that might not happen.
Either way, you'll still be installing module-federation/enhanced.
I need to be able to release new versions of the system within an SLA of 15 minutes; getting blocked by release cycles is a non-negotiable.
The challenge we have seen is Bytedance engineering moves faster than any community based project. Federation is a core technology that powers most of the thousands of apps we own.
Backmerging parts is a spare time effort of which only i do.
Just migrate.

**Comment by ScriptedAlchemy** [comment]
If there is it'll be here. https://module-federation.io/

But i mean, you install the plugin, replace the import from webpack to enhanced for federation and that's it.

**Comment by einarq** [comment]
I’ll give it go, thanks!On 12 Nov 2024, at 21:02, Zack Jackson ***@***.***> wrote:﻿
If there is it'll be here. https://module-federation.io/
But i mean, you install the plugin, replace the import from webpack to enhanced for federation and that's it.

**Comment by einarq** [comment]
Getting this error on a fairly simple project:

[image: image.png]

I know I should try to create a repro :) Just thinking that maybe there is
a common cause of this?

**Comment by ScriptedAlchemy** [comment]
Stop using email and send these issues to my core repo instead. No image response here.

**Comment by einarq** [comment]
https://github.com/module-federation/core/discussions/3281

**Comment by alexander-akait** [comment]
bump

**Comment by LeonxLJX** [comment]
I'd like to take this one.

I've read through the report. Before changing anything I'd like to reproduce it against current `main` first — my reasoning is that a fix anchored to the actual observed behavior is much easier to review than one inferred from the report alone, especially if the surrounding code has moved since this was filed.

Plan: reproduce and confirm the root cause, then make the smallest change that addresses it (plus a regression test if the repo has coverage for this area). If the root cause turns out to live somewhere other than where the report points, I'll report back here before forcing a change.

Anything already known about this that I should read first?

(claiming via @LeonxLJX)
